core/src/apps/cardano/layout.py

```diff
@@ -276,32 +281,92 @@ async def confirm_inline_datum(first_chunk: bytes, inline_datum_size: int) -> No
 async def confirm_reference_script(
     first_chunk: bytes, reference_script_size: int
 ) -> None:
-    await _confirm_data_chunk(
+    await _confirm_tx_data_chunk(
         "confirm_reference_script",
         "Reference script",
         first_chunk,
         reference_script_size,
     )
 
 
-async def _confirm_data_chunk(
-    br_type: str, title: str, first_chunk: bytes, data_size: int
+async def confirm_message_payload(
+    payload_first_chunk: bytes,
+    payload_hash: bytes,
+    payload_size: int,
+    is_signing_hash: bool,
+    display_ascii: bool,
 ) -> None:
-    MAX_DISPLAYED_SIZE = 56
-    displayed_bytes = first_chunk[:MAX_DISPLAYED_SIZE]
+    props: list[PropertyType]
+
+    max_displayed_bytes = (
+        _DEFAULT_MAX_DISPLAYED_CHUNK_SIZE if is_signing_hash else MAX_CHUNK_SIZE
+    )
+
+    if not payload_first_chunk:
+        assert payload_size == 0
+        props = _get_data_chunk_props(
+            title="Empty message",
+            first_chunk=payload_first_chunk,
+            data_size=payload_size,
+        )
+    elif display_ascii:
+        if not is_unambiguous_ascii(payload_first_chunk):
+            raise ProcessError(
+                "Payload cannot be decoded as ASCII or its decoding leads to a visually ambiguous string"
+            )
+        props = _get_data_chunk_props(
+            title="Message text",
+            first_chunk=payload_first_chunk,
+            data_size=payload_size,
+            max_displayed_size=max_displayed_bytes,
+            decoder=lambda chunk: chunk.decode("ascii"),
+        )
+    else:
+        props = _get_data_chunk_props(
+            title="Message hex",
+            first_chunk=payload_first_chunk,
+            data_size=payload_size,
+            max_displayed_size=max_displayed_bytes,
+        )
+
+    props.append(("Message hash:", payload_hash))
+
+    await confirm_properties(
+        "confirm_message_payload",
+        title="Confirm message",
+        props=props,
+        br_code=BRT_Other,
+    )
+
+
+def _get_data_chunk_props(
+    title: str,
+    first_chunk: bytes,
+    data_size: int,
+    max_displayed_size: int = _DEFAULT_MAX_DISPLAYED_CHUNK_SIZE,
+    decoder: Callable[[bytes], bytes | str] | None = None,
+) -> list[PropertyType]:
+    displayed_bytes = first_chunk[:max_displayed_size]
     bytes_optional_plural = "byte" if data_size == 1 else "bytes"
-    props: list[tuple[str, bytes | None]] = [
+    props: list[PropertyType] = [
         (
             f"{title} ({data_size} {bytes_optional_plural}):",
-            displayed_bytes,
+            decoder(displayed_bytes) if decoder else displayed_bytes,
         )
     ]
-    if data_size > MAX_DISPLAYED_SIZE:
+    if data_size > max_displayed_size:
         props.append(("...", None))
```

Comment:
Not sure I understand what happens here, but should the `"..."` be another prop rather than be appended to the value of the truncated chunk?

PS: I see this was here before, so it is probably fine... or could/should it be improved?